### ISSUE TITLE
API Access

### ISSUE THREAD

**Issue by ChiperSoft** [bigboringsystem/bigboringsystem#196]
*Followup from in-chat discussion, posted here for more in-depth discussion of implementation.*

Currently there's little to no proper api integration available on BBS. All of the routes are coded for access purely from a browser session, and the only way to send data to BBS is to create a session through the sms login and then use the session cookie.

To facilitate external services, there needs to be:

1. An authentication method that doesn't use SMS (managed keys, oauth, something)
2. API endpoints that return and accept JSON instead of html and form data. This could be done on existing endpoints via header detection.

Thoughts? Ideas? Concerns?

**Comment by TheAlphaNerd** [comment]
I think that this is a decent idea... although we should definitely consider the scope that this could create.

Specifically if the back end of BBS exposes a sane API it raises the question if the back end should be responsible for serving the front end.  The client application is fairly thin.  If something like a proper API and Auth were to land it would probably make sense to completely decouple the two applications.

**Comment by imakewebthings** [comment]
If somebody could chime in about what concrete uses and applications one would use the API for if it were available, I'd love to hear more.

**Comment by ChiperSoft** [comment]
I initially started this conversation because I was brainstorming adding an IRC proxy for the chat system, so that I could have the BBS chat running in my IRC client instead of having to keep a browser window open.

As the chat is currently working, I could create an account to connect to the site with and receive chat events to broadcast on the irc server, but that's the limit of what it could do. There's no way for me to authenticate people who connect to the IRC server against BBS so that I can confirm they're allowed to even be on the server, so I started brainstorming methods that could be achieved.

Using an API key as the irc password seemed like a good solution, similar to how Gitter's IRC proxy works.

There's still a mountain of other limitations in the way of implementing sending messages from the proxy, but I figured one problem at a time. (In theory, once authentication is solved, I could open a socket.io link for each user that connected to the server and use that link to relay their messages. It'd be really inefficient, but it would work.)
